File: backend/app/schemas/scenario.py
"""Complete scenario and request/response schemas."""

from __future__ import annotations

from typing import Any

from pydantic import Field, model_validator

from .actors import (
    Actor,
    CyclistActor,
    EgoVehicle,
    PedestrianActor,
    VehicleActor,
)
from .common import Assumption, NonNegFloat, PositiveFloat, StrictModel, Unknown
from .mutations import MutationSpec
from .oracles import OracleResult, SafetyOracle
from .road import RoadLayout
from .triggers import Trigger


class ScenarioSpec(StrictModel):
    """Complete Stage-1 scenario specification (SI units, deterministic)."""

    id: str
    name: str
    description: str = ""
    duration_s: PositiveFloat = Field(..., description="Simulation horizon [s]")
    timestep_s: PositiveFloat = Field(
        default=0.1, description="Fixed integration step [s]; must be 0.1 for Stage 1"
    )
    road: RoadLayout
    ego: EgoVehicle
    actors: list[VehicleActor | CyclistActor | PedestrianActor] = Field(
        default_factory=list
    )
    triggers: list[Trigger] = Field(default_factory=list)
    oracles: list[SafetyOracle] = Field(default_factory=list)
    assumptions: list[Assumption] = Field(default_factory=list)
    unknowns: list[Unknown] = Field(default_factory=list)
    mutation: MutationSpec | None = None

    @model_validator(mode="after")
    def _check_timestep(self) -> ScenarioSpec:
        if abs(self.timestep_s - 0.1) > 1e-12:
            raise ValueError("Stage 1 requires timestep_s == 0.1")
        return self

    def all_actors(self) -> list[Actor]:
        return [self.ego, *self.actors]


class ValidationIssue(StrictModel):
    code: str
    message: str
    path: str | None = None


class ValidateRequest(StrictModel):
    scenario: ScenarioSpec


class ValidateResponse(StrictModel):
    valid: bool
    issues: list[ValidationIssue] = Field(default_factory=list)


class SimulateRequest(StrictModel):
    scenario: ScenarioSpec


class ActorFrameState(StrictModel):
    id: str
    x: float
    y: float
    vx: float
    vy: float
    heading_deg: float
    length: float
    width: float
    actor_type: str


class SimulationFrame(StrictModel):
    t: NonNegFloat
    actors: list[ActorFrameState]
    collisions: list[tuple[str, str]] = Field(default_factory=list)
    ego_speed: NonNegFloat
    min_ttc: float | None = None


class SimulationMetrics(StrictModel):
    duration_s: float
    timestep_s: float
    frame_count: int
    collision_count: int
    min_ttc: float | None
    max_acceleration: float
    max_jerk: float
    lane_boundary_violations: int
    initial_overlap: bool
    oracle_results: list[OracleResult]


class SimulateResponse(StrictModel):
    scenario_id: str
    valid: bool
    validation_issues: list[ValidationIssue] = Field(default_factory=list)
    frames: list[SimulationFrame] = Field(default_factory=list)
    metrics: SimulationMetrics | None = None
    trajectories: dict[str, list[tuple[float, float]]] = Field(default_factory=dict)


class PresetSummary(StrictModel):
    id: str
    name: str
    description: str
    default_testing_goal: str = ""
    kind: str = Field(
        default="scenario",
        description="scenario | impossible  -  demo classification for the UI",
    )


class HealthResponse(StrictModel):
    status: str
    service: str
    version: str
    deterministic: bool


class CompileRequest(StrictModel):
    """Compile a seed scene + natural-language testing goal via OpenAI (server-side)."""

    seed_scene: ScenarioSpec
    testing_goal: str = Field(..., min_length=1)
    run_simulation: bool = True


class CompileResponse(StrictModel):
    mode: str
    model: str | None = None
    ok: bool
    error_code: str | None = None
    error: str | None = None
    target_kind: str | None = None
    json_parse_ok: bool = False
    schema_valid: bool = False
    physical_valid: bool = False
    parsed: dict[str, Any] | None = None
    validation_issues: list[ValidationIssue] = Field(default_factory=list)
    simulation: SimulateResponse | None = None
    latency_s: float | None = None
    usage: dict[str, Any] | None = None
    # raw_text intentionally omitted from public API to keep payloads smaller;
    # available in offline eval artifacts only.

File: frontend/src/types/scenario.ts
export interface Position2D {
  x: number;
  y: number;
}

export interface Velocity2D {
  vx: number;
  vy: number;
}

export interface Dimensions2D {
  length: number;
  width: number;
}

export interface ActorBehavior {
  type: string;
  trigger_id?: string | null;
  post_trigger_velocity?: Velocity2D | null;
  target_lane_id?: string | null;
  lateral_speed?: number | null;
}

export interface ActorState {
  id: string;
  actor_type: string;
  position: Position2D;
  velocity: Velocity2D;
  dimensions: Dimensions2D;
  lane_id?: string | null;
  heading_deg: number;
  behavior: ActorBehavior;
}

export interface Lane {
  id: string;
  center_y: number;
  width: number;
  direction: 1 | -1;
}

export interface RoadLayout {
  kind: string;
  length: number;
  lanes: Lane[];
  intersection_center?: [number, number] | null;
  intersection_size?: number | null;
  cross_lane_width?: number;
}

export interface Trigger {
  id: string;
  type: string;
  time_s?: number | null;
  distance_m?: number | null;
  reference_point?: Position2D | null;
  region_half_extent_m?: number | null;
}

export interface SafetyOracle {
  id: string;
  type: string;
  threshold?: number | null;
  actor_id?: string | null;
}

export interface ScenarioSpec {
  id: string;
  name: string;
  description: string;
  duration_s: number;
  timestep_s: number;
  road: RoadLayout;
  ego: ActorState;
  actors: ActorState[];
  triggers: Trigger[];
  oracles: SafetyOracle[];
  assumptions?: { id: string; statement: string }[];
  unknowns?: { id: string; statement: string }[];
}

export interface PresetSummary {
  id: string;
  name: string;
  description: string;
}

export interface ValidationIssue {
  code: string;
  message: string;
  path?: string | null;
}

export interface ActorFrameState {
  id: string;
  x: number;
  y: number;
  vx: number;
  vy: number;
  heading_deg: number;
  length: number;
  width: number;
  actor_type: string;
}

export interface SimulationFrame {
  t: number;
  actors: ActorFrameState[];
  collisions: [string, string][];
  ego_speed: number;
  min_ttc: number | null;
}

export interface OracleResult {
  id: string;
  type: string;
  passed: boolean;
  value?: number | null;
  message: string;
}

export interface SimulationMetrics {
  duration_s: number;
  timestep_s: number;
  frame_count: number;
  collision_count: number;
  min_ttc: number | null;
  max_acceleration: number;
  max_jerk: number;
  lane_boundary_violations: number;
  initial_overlap: boolean;
  oracle_results: OracleResult[];
}

export interface SimulateResponse {
  scenario_id: string;
  valid: boolean;
  validation_issues: ValidationIssue[];
  frames: SimulationFrame[];
  metrics: SimulationMetrics | null;
  trajectories: Record<string, [number, number][]>;
}
